### PR TITLE
Sharded SDPAFwdOp

### PR DESCRIPTION
**Pull request by cowanmeg**
Temporarily add support for a sharded forward scaled dot product attention.
Currently, we only support DID parallelization on the logical domain which requires us to split and parallelize an axis at the logical level see https://github.com/NVIDIA/Fuser/issues/2563. This is a hack until we support DID parallelization on the loop domain after which this PR can be reverted.

Restrictions:
1. q,k,v inputs are manually sharded _before_ the SDPAFwdOp is created. We cannot rely on sharding propagation or sharding after the Fusion is created, because the dimension checks are called when the op is created. 
2. Only the head dimension is sharded and all inputs and outputs have either a sharded head dimension or unshaded. 
3. DID axis is the outermost axis. This is because during evaluation if we see 5 dimensions, it is assumed the first is the DID axis and is appropriately squeezed from the inputs and unsqueezed onto the outputs.

### CODE REVIEW

File: csrc/ir/nodes.cpp
Comment (Priya2698) [inline csrc/ir/nodes.cpp:4399]: This should be after L4401-4402 since dim=3 is assumed to be last dimension. Or `output.sizes()[3]` should be changed to `output.sizes()[-1]`

File: csrc/ir/nodes.cpp
Comment (Priya2698) [inline csrc/ir/nodes.cpp:4397]: Why do we need the DID axis in `log_sumexp`?

Comment (cowanmeg) [reply csrc/ir/nodes.cpp:4397]: Good question, we only need it for the backwards, so we don't need to track the DID axis. I'll remove it!

File: csrc/root_domain_map.cpp
Comment (wujingyue) [inline csrc/root_domain_map.cpp:241]:
```suggestion
    size_t num_device_dim = producer_logical.at(0)->isDeviceDim() ? 1 : 0;
```

It's better for num_device_dim and idx to have the same type so comparisons are unambiguous.

File: csrc/root_domain_map.cpp
Comment (wujingyue) [inline csrc/root_domain_map.cpp:259]: Still needed given https://github.com/NVIDIA/Fuser/pull/2565/files#r1672758030?

**Review by wujingyue — APPROVED** [review]
Nice! I'd wait for Priya's approval.

**Comment by cowanmeg** [comment]
!build

**Review by Priya2698 — APPROVED** [review]
LGTM.

**Comment by cowanmeg** [comment]
Note: failing tests are unrelated and caused by `'nlohmann/json.hpp' file not found`

**Pull request merged**
